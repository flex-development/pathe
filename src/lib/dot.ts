/**
 * @file dot
 * @module pathe/lib/dot
 */

import type { Dot } from '@flex-development/pathe'

/**
 * Dot character (`'.'`).
 *
 * @see {@linkcode Dot}
 *
 * @category
 *  utils
 *
 * @const {Dot} dot
 */
const dot: Dot = '.'

export default dot
