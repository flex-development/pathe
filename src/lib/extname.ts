/**
 * @file extname
 * @module pathe/lib/extname
 */

import { DOT } from '#src/internal/constants'
import ensurePosix from '#src/internal/ensure-posix'
import isDrivePath from '#src/internal/is-drive-path'
import isSep from '#src/internal/is-sep'
import validateString from '#src/internal/validate-string'
import type { Ext } from '#src/types'
import type { EmptyString } from '@flex-development/tutils'

/**
 * Returns the extension of the given `path`, from the last occurrence of the
 * `.` (dot) character to end of the string in the last portion of the path.
 *
 * If there is no `.` in the last portion of the path, or if there are no `.`
 * characters other than the first character of the path's [`basename`][1], an
 * empty string will be returned.
 *
 * [1]: {@link ./basename.ts}
 *
 * @param {string} path - Path to evaluate
 * @return {EmptyString | Ext} Extension of `path` or empty string
 * @throws {TypeError} If `path` is not a string
 */
const extname = (path: string): EmptyString | Ext => {
  validateString(path, 'path')

  // exit early if path does not contain any dot characters
  if (!path.includes(DOT)) return ''

  // ensure path meets posix standards
  path = ensurePosix(path)

  /**
   * Index to begin searching for extension.
   *
   * @var {number} offset
   */
  let offset: number = 0

  /**
   * Start index of {@linkcode path}'s basename.
   *
   * @var {number} part
   */
  let part: number = 0

  /**
   * State of characters, if any, before first dot character and after any path
   * separators in {@linkcode path}.
   *
   * @var {number} predot
   */
  let predot: number = 0

  /**
   * Directory separator match check.
   *
   * @var {boolean} sep_match
   */
  let sep_match: boolean = true

  /**
   * Start index of extension.
   *
   * @var {number} start
   */
  let start: number = -1

  /**
   * End index of extension.
   *
   * @var {number} end
   */
  let end: number = -1

  // check for drive path so as not to mistake the following path separator as
  // an extra separator at the end of the path that can be disregarded
  if (path.length >= 2 && isDrivePath(path)) {
    offset = part = 2
  }

  // get start and end indices of extension
  for (let i = path.length - 1; i >= offset; --i) {
    /**
     * Character at {@linkcode i} in {@linkcode path}.
     *
     * @const {string} char
     */
    const char: string = path.charAt(i)

    // adjust start index of basename
    if (isSep(char)) {
      if (!sep_match) {
        // encountered separator that is not trailing separator
        part = i + 1
        break
      }

      continue
    }

    // set end index of extension
    if (end === -1) {
      end = i + 1
      sep_match = false
    }

    // set start index of extension and/or update predot state
    if (char === DOT) {
      // set start index of extension
      if (start === -1) start = i
      else if (predot !== 1) predot = 1
    } else if (start !== -1) {
      // a non-dot and non-separator character was encountered before the dot
      // character, so there is a good chance at having a non-empty extension
      predot = -1
    }
  }

  return start === -1 || end === -1 || predot === 0
    ? ''
    : start === part + 1 && start === end - 1 && predot === 1
    ? ''
    : (path.slice(start, end) as Ext)
}

export default extname
