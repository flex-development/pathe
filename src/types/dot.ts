/**
 * @file Type Aliases - Dot
 * @module pathe/types/Dot
 */

/**
 * A dot character (`'.'`).
 */
type Dot = '.'

export type { Dot as default }
