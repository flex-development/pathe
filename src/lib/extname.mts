/**
 * @file extname
 * @module pathe/lib/extname
 */

import { DRIVE_PATH_REGEX } from '#internal/constants'
import validateURLString from '#internal/validate-url-string'
import type basename from '#lib/basename'
import dot from '#lib/dot'
import isSep from '#lib/is-sep'
import toPath from '#lib/to-path'
import toPosix from '#lib/to-posix'
import type { EmptyString, Ext } from '@flex-development/pathe'

/**
 * Get the file extension of `input` from the last occurrence of the `.` (dot)
 * character (`.`) to end of the string in the last portion of `input`.
 *
 * If there is no `.` in the last portion of `input`, or if there are no `.`
 * characters other than the first character of the {@linkcode basename} of
 * `input`, an empty string is returned.
 *
 * > ðŸ‘‰ **Note**: If `input` is a {@linkcode URL}, or can be parsed to a `URL`,
 * > it will be converted to a path using {@linkcode toPath}.
 *
 * @see {@linkcode EmptyString}
 * @see {@linkcode Ext}
 *
 * @category
 *  core
 *
 * @this {void}
 *
 * @param {URL | string} input
 *  The {@linkcode URL}, URL string, or path to handle
 * @return {EmptyString | Ext}
 *  Extension of `input` or empty string
 */
function extname(this: void, input: URL | string): EmptyString | Ext {
  validateURLString(input, 'input')
  input = toPosix(toPath(input))

  if (!input.includes(dot)) return ''

  /**
   * Index to begin searching for extension.
   *
   * @var {number} offset
   */
  let offset: number = 0

  /**
   * Start index of {@linkcode input}'s basename.
   *
   * @var {number} part
   */
  let part: number = 0

  /**
   * State of characters, if any, before first dot character and after any path
   * separators in {@linkcode input}.
   *
   * @var {number} predot
   */
  let predot: number = 0

  /**
   * Boolean indicating a path separator was seen.
   *
   * @var {boolean} separator
   */
  let separator: boolean = true

  /**
   * Start index of extension.
   *
   * @var {number} start
   */
  let start: number = -1

  /**
   * End index of extension.
   *
   * @var {number} end
   */
  let end: number = -1

  // check for drive path so as not to mistake the next path separator as an
  // extra separator at the end of the path that can be disregarded
  if (input.length >= 2 && DRIVE_PATH_REGEX.test(input)) {
    offset = part = 2
  }

  // get start and end indices of extension
  for (let i = input.length - 1; i >= offset; --i) {
    /**
     * Current character in {@linkcode input}.
     *
     * @const {string} char
     */
    const char: string = input[i]!

    if (isSep(char)) {
      if (!separator) {
        // stop if a non-trailing path separator was reached
        part = i + 1
        break
      }

      continue
    }

    if (end === -1) {
      // reached first non-path separator -> end of extension
      end = i + 1
      separator = false
    }

    if (char === dot) {
      // first dot -> start of extension
      if (start === -1) start = i
      else if (predot !== 1) predot = 1
    } else if (start !== -1) {
      // a non-dot and non-separator character was encountered before the dot
      // character, so there is a good chance at having a non-empty extension
      predot = -1
    }
  }

  if (
    start === -1 ||
    end === -1 ||
    // non-dot character immediately before the dot
    predot === 0 ||
    // right-most trimmed path component is exactly '..'
    (predot === 1 && start === end - 1 && start === part + 1)
  ) {
    return ''
  }

  return input.slice(start, end) as Ext
}

export default extname
