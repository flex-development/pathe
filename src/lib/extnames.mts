/**
 * @file extnames
 * @module pathe/lib/extnames
 */

import validateURLString from '#internal/validate-url-string'
import extname from '#lib/extname'
import type { EmptyString, Ext } from '@flex-development/pathe'

/**
 * Get a list of file extensions for `input`.
 *
 * @see {@linkcode Ext}
 * @see {@linkcode extname}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {URL | string} input
 *  The {@linkcode URL}, URL string, or path to handle
 * @return {Ext[]}
 *  List of extensions
 */
function extnames(this: void, input: URL | string): Ext[] {
  validateURLString(input, 'input')

  /**
   * List of extensions.
   *
   * @const {Ext[]} extensions
   */
  const extensions: Ext[] = []

  /**
   * Current path.
   *
   * @var {string} subpath
   */
  let subpath: string = String(input)

  if (subpath) {
    while (true) {
      /**
       * Current extension.
       *
       * @const {EmptyString | Ext} ext
       */
      const ext: EmptyString | Ext = extname(subpath)

      if (ext === '') break

      extensions.unshift(ext)
      subpath = subpath.slice(0, subpath.lastIndexOf(ext))
    }
  }

  return extensions
}

export default extnames
