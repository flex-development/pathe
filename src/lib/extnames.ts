/**
 * @file extnames
 * @module pathe/lib/extnames
 */

import validateString from '#internal/validate-string'
import dot from '#lib/dot'
import extname from '#lib/extname'
import toPosix from '#lib/to-posix'
import type { EmptyString, Ext } from '@flex-development/pathe'

/**
 * Get a list of file extensions for `path`.
 *
 * @see {@linkcode Ext}
 * @see {@linkcode extname}
 *
 * @param {string} path
 *  Path to handle
 * @return {Ext[]}
 *  List of extensions
 */
function extnames(path: string): Ext[] {
  validateString(path, 'path')

  /**
   * List of extensions.
   *
   * @const {Ext[]} extensions
   */
  const extensions: Ext[] = []

  /**
   * Current path.
   *
   * @var {string} subpath
   */
  let subpath: string = toPosix(path)

  while (subpath.includes(dot)) {
    /**
     * Current extension.
     *
     * @const {EmptyString | Ext} ext
     */
    const ext: EmptyString | Ext = extname(subpath)

    if (ext === '') break

    extensions.unshift(ext)
    subpath = subpath.slice(0, subpath.lastIndexOf(ext))
  }

  return extensions
}

export default extnames
