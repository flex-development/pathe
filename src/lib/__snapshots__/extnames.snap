// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/extnames > should return list of extensions ("") 1`] = `[]`;

exports[`unit:lib/extnames > should return list of extensions (".") 1`] = `[]`;

exports[`unit:lib/extnames > should return list of extensions (".remarkignore") 1`] = `[]`;

exports[`unit:lib/extnames > should return list of extensions (".remarkrc.mjs") 1`] = `
[
  ".mjs",
]
`;

exports[`unit:lib/extnames > should return list of extensions ("/") 1`] = `[]`;

exports[`unit:lib/extnames > should return list of extensions ("eslint.base.config.mjs") 1`] = `
[
  ".base",
  ".config",
  ".mjs",
]
`;

exports[`unit:lib/extnames > should return list of extensions ("grease.config.mjs") 1`] = `
[
  ".config",
  ".mjs",
]
`;

exports[`unit:lib/extnames > should return list of extensions ("src/lib/extnames.ts") 1`] = `
[
  ".ts",
]
`;
